lib/vector-buffers/src/variants/disk_v1/mod.rs

```diff
@@ -241,47 +250,85 @@ where
     // key K, then the next key we generate will be K+10.  This lets us take the difference between
     // the last key and the first key to get the number of actual events in the buffer, minus the
     // events contained in the last key/value pair.  We decode it below to get that, too.
-    let mut item_size = last_key.unwrap_or(0) as u64 - first_key.unwrap_or(0) as u64;
+    state.total_events = state.last_key.unwrap_or(0) as u64 - state.first_key.unwrap_or(0) as u64;
 
-    if last_key.is_some() {
+    if let Some(last_key) = state.last_key.as_ref().copied() {
         let iter = db.iter(ReadOptions::new());
         iter.seek_to_last();
         if iter.valid() {
             let val = iter.value();
-            if let Ok(item) = T::decode(T::get_metadata(), &val[..]) {
-                item_size += item.event_count() as u64;
+            match T::decode(T::get_metadata(), &val[..]) {
+                Ok(record) => {
+                    let event_count = record.event_count();
+                    state.total_events += event_count as u64;
+                    state.write_offset = Some(last_key.wrapping_add(event_count));
+                }
+                Err(e) => {
+                    // If the last record couldn't be decoded, we know the reader is never going to
+                    // do anything with this read to begin with, so we make the conscious decision
+                    // to delete it here and now.
+                    //
+                    // If we don't delete it now, we have to compensate in the reader code when we
+                    // hit and make sure we don't count it as a true event for the purposes of
+                    // metrics, but do properly delete it, and so on.  We know we have to delete it,
+                    // so if we just do that now, then our reader logic can stay a bit cleaner.
+                    //
+                    // We also have to delete it now, rather than just setting the write offset to
+                    // overwrite it, because otherwise, the reader might get to it before a write
+                    // comes in that overwrites it.
+                    error!(
+                        decode_error = %e,
+                        "Detected undecodable record when querying initial state of buffer. Dropping record and continuing."
+                    );
+
+                    // Set our write offset, and update the last key, to compensate for the fact
+                    // this record is going away.  Also update our buffer byte size value, too.
+                    state.write_offset = Some(last_key);
+                    if state.first_key != state.last_key {
+                        // If this isn't the only record, then go back by one to get the end of the
+                        // previous record.  Otherwise, our last key shouldn't move at all.
+                        state.last_key = Some(last_key.wrapping_sub(1));
```

Comment:
Will this still work if the previous record contains more than one event?

Reply:
Actually, thinking this through: no, what I'm doing there doesn't make a ton of sense.  We also don't need to update `last_key`, though, since it's really just a temporary for the sake of calculating the total event count.

I'll change this section up to reflect that all we really care about is the size in events/bytes, and the read/write offset.

Reply:
Addressed in 6da55f3cf.